tools/deploy-cli/deploy-cli.go

```diff
@@ -0,0 +1,324 @@
+		kubeconfigPath = os.Getenv("KUBECONFIG_PATH")
+		if kubeconfigPath == "" {
+			homeDir, err := os.UserHomeDir()
+			if err != nil {
+				return "", fmt.Errorf("failed to get user home directory: %v", err)
+			}
+			kubeconfigPath = filepath.Join(homeDir, ".kube", "config")
+		}
+	}
+
+	// Verify the file exists and is readable
+	_, err := os.Stat(kubeconfigPath)
+	if os.IsNotExist(err) || err != nil {
+		return "", fmt.Errorf("specified kubeconfig file does not exist or is not readable: %s", kubeconfigPath)
+	}
+
+	return kubeconfigPath, nil
+}
+
+// generateClusterProxy creates a new ClusterProxy instance using the provided kubeconfig path.
+func generateClusterProxy(kubeconfigPath string) (*framework.ClusterProxy, error) {
+	scheme := runtime.NewScheme()
+	framework.TryAddDefaultSchemes(scheme)
+	clusterProxy := framework.NewClusterProxy("deploy-cli", kubeconfigPath, scheme)
+	if clusterProxy == nil {
+		return nil, fmt.Errorf("failed to create cluster proxy")
+	}
+	return &clusterProxy, nil
+}
+
+// deployWithKustomizeAndApply first generates the YAML configuration by running Kustomize build on the overlay directory,
+// then outputs the generated YAML to a temp file within the overlay dir, and then applies
+// the configuration to the Kubernetes cluster specified in the kubeconfig file.
+func deployWithKustomizeAndApply(overlayPath string) error {
+	yamlOutput, err := BuildKustomizeManifest(overlayPath)
+
+	if err != nil {
+		return fmt.Errorf("failed to apply YAML: %v", err)
+	}
+
+	kubeconfigPath, err := getKubeconfigPath()
+	if err != nil {
+		return fmt.Errorf("failed to apply YAML: %v", err)
+	}
+
+	ctx := context.Background()
+	if err := testexec.KubectlApply(ctx, kubeconfigPath, yamlOutput); err != nil {
+		return fmt.Errorf("failed to apply YAML: %v", err)
+	}
+
+	return nil
+}
+
+// cleanup removes temporary files created for basic auth credentials during deployment.
+func cleanup(deployBasicAuthFlag, deployBMOFlag, deployIronicFlag bool, tempBMOOverlay, tempIronicOverlay string) {
+	if deployBasicAuthFlag {
+		if deployBMOFlag {
+			os.Remove(filepath.Join(tempBMOOverlay, "ironic-username"))
+			os.Remove(filepath.Join(tempBMOOverlay, "ironic-password"))
+			os.Remove(filepath.Join(tempBMOOverlay, "ironic-inspector-username"))
+			os.Remove(filepath.Join(tempBMOOverlay, "ironic-inspector-password"))
+		}
+
+		if deployIronicFlag {
+			os.Remove(filepath.Join(tempIronicOverlay, "ironic-auth-config"))
+			os.Remove(filepath.Join(tempIronicOverlay, "ironic-inspector-auth-config"))
+			os.Remove(filepath.Join(tempIronicOverlay, "ironic-htpasswd"))
+			os.Remove(filepath.Join(tempIronicOverlay, "ironic-inspector-htpasswd"))
+		}
+	}
+}
+
+func usage() {
+	fmt.Println(`Usage : deploy [options]
+Options:
+	-h:	show this help message
+	-b:	deploy BMO
+	-i:	deploy Ironic
+	-t:	deploy with TLS enabled
+	-n:	deploy without authentication
+	-k:	deploy with keepalived
+	-m:	deploy with mariadb (requires TLS enabled)`)
+}
```

Comment:
Please add at least required environment variables (IRONIC_HOST_IP), preferably also optional variables.

Reply:
Hmm. What are those required variables?
IMO there are just too many of them, we shouldn't make them flags. In this version, we take them from the environment, just like `deploy.sh` does. In later improvement, we could add a config file, so that user could provide their own vars.

Reply:
I don't mean to add flags for them, I just mean that we should mention at least the required environment variables. If I'm not mistaken it will immediately exit with an error if you don't set IRONIC_HOST_IP.

Reply:
Yes, then we're on the same page. I'm just having a hard time deciding which env vars should be required xD. 
My initial intention was that we will gradually add those required variables as we see fit, but of course we could start with something now. :-?

Reply:
I think it would be as simple as this: If you can run the command successfully without the env var then it is not required. So try running it and see what would be needed. I noticed [this place](https://github.com/metal3-io/baremetal-operator/blob/ed781396e524255850e4ddc54c27e2cdcd7079eb/tools/deploy-cli/deploy-cli.go#L190-L192) where we fail if the variable is empty. There could be more that I missed though.